manager/state/raft/transport/transport.go

```diff
@@ -0,0 +1,305 @@
+package transport
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/net/context"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/coreos/etcd/raft"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/docker/swarmkit/api"
+	"github.com/docker/swarmkit/log"
+	"github.com/pkg/errors"
+)
+
+// ErrIsNotFound indicates that peer was never added to transport.
+var ErrIsNotFound = errors.New("peer not found")
+
+// Raft is interface which represents Raft API for transport package.
+type Raft interface {
+	ReportUnreachable(id uint64)
+	ReportSnapshot(id uint64, status raft.SnapshotStatus)
+	IsIDRemoved(id uint64) bool
+
+	NodeRemoved()
+}
+
+// Config for Transport
+type Config struct {
+	SendTimeout time.Duration
+	Credentials credentials.TransportCredentials
+
+	Raft Raft
+}
+
+// Transport is structure which manages remote raft peers and sends messages
+// to them.
+type Transport struct {
+	config *Config
+
+	unknownc chan raftpb.Message
+
+	mu      sync.Mutex
+	peers   map[uint64]*peer
+	stopped bool
+
+	ctx  context.Context
+	done chan struct{}
+}
+
+// New returns new Transport with specified Context and Config.
+// Transport will shutdown on context cancel.
+func New(ctx context.Context, cfg *Config) *Transport {
+	t := &Transport{
+		peers:    make(map[uint64]*peer),
+		config:   cfg,
+		unknownc: make(chan raftpb.Message),
+		done:     make(chan struct{}),
+		ctx:      ctx,
+	}
+	go t.run(ctx)
+	go t.unknownSender(ctx)
+	return t
+}
+
+func (t *Transport) run(ctx context.Context) {
+	<-ctx.Done()
+	log.G(ctx).Debug("stop transport")
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	t.stopped = true
+	for _, p := range t.peers {
+		p.cancel()
+		<-p.done
+	}
+	close(t.done)
+}
+
+func (t *Transport) sendUnknownMessage(ctx context.Context, m raftpb.Message) error {
+	ctx, cancel := context.WithTimeout(ctx, 2*t.config.SendTimeout)
+	defer cancel()
+	p, err := t.resolvePeer(ctx, m.To)
+	if err != nil {
+		return errors.Wrapf(err, "failed to resolve peer")
+	}
+	defer p.cancel()
+	if err := p.sendProcessMessage(ctx, m); err != nil {
+		return errors.Wrapf(err, "failed to send message")
+	}
+	return nil
+}
+
+// unknownSender sends messages to unknown peers. It creates new peer for each
+// message and discards it after send.
+func (t *Transport) unknownSender(ctx context.Context) {
```

Comment:
I don't understand why we need to be sure that the raft is up in this case. `ReportUnreachable` and `ReportSnapshot` is no-op in those cases. Cancelling request should be enough.